### PR TITLE
[WIP] tests: add comprehensive memory limits testing framework (Closes #4247)

### LINKED ISSUE #4247: Test VM memory limits with maximum RAM and heavy workload
We need to create a test that runs a VM with the maximum available RAM for the system and performs a heavy storage and network workload within the VM. The purpose of this test is to validate that our memory overhead estimations and the memory limits set for the VM are accurate and stable under maximum resource utilization.

For example, for a device with 4Gb of RAM, we know that we can run a VM with 2.1GB of RAM, as the rest goes to the EVE system itself and the VMM overhead. But we need to check these estimations.

We can implement it in any suitable way and try to integrate it later.

### PR DESCRIPTION
**Pull request by jeff-zed**
## Description

This PR adds a testing framework to validate EVE's memory management capabilities, specifically addressing VM memory limits with maximum RAM allocation as described in issue #4247.

The framework includes:
- **Memory allocation testing**: Tests containers/VMs with maximum possible RAM allocation up to system limits
- **Memory overhead validation**: Measures and validates EVE system overhead vs. available container memory
- **Boundary detection**: Binary search algorithm to find exact memory allocation limits
- **Stress testing**: Validates behavior under memory pressure and concurrent allocations
- **OOM handling**: Tests graceful handling of over-allocation attempts

The test suite can be run standalone or integrated into CI/CD pipelines, providing automated validation of EVE's memory management and resource constraints.

**Fixes:** #4247

## How to test and validate this PR

```bash
# Clone and build EVE

# Clone and build Eden

# Run quick validation (5 minutes)
.tests/memory-testing/scripts/simple-memory-test.sh

# Run comprehensive memory tests
.tests/memory-testing/scripts/maximum-memory-test.sh --all-tests

# Find exact memory allocation boundary
.tests/memory-testing/scripts/boundary-finder.sh --start 3066 --end 3072 --precision 1
```

### Expected Results:
- Memory allocation respects system limits
- EVE properly rejects over-allocation attempts
- System remains stable under memory pressure
- Clear error messages for allocation failures

```bash
# CSV of tests
.tests/memory-testing/results/<test>/<test>.csv
# Summary
.tests/memory-testing/results/<test>/<test>.md
```

## Changelog notes

Added memory limits testing framework for validating VM/container memory allocation, overhead calculations, and resource constraint handling. Includes automated tests for memory boundaries, stress scenarios, and concurrent allocations.

## PR Backports

* 14.5-stable: Not required (testing framework addition)
* 13.4-stable: Not required (testing framework addition)

## Checklist

- [x] I've provided a proper description
- [x] I've added the proper documentation
- [ ] I've tested my PR on amd64 device
- [ ] I've tested my PR on arm64 device
- [x] I've written the test verification instructions
- [x] I've set the proper labels to this PR

I do not have an amd64 device or a "standard" (non-Apple) arm64 device for testing.

And the last but not least:
- [x] I've checked the boxes above, or I've provided a good reason why I didn't check them.

### CODE REVIEW

**Comment by OhmSpectator** [comment]
Wow. We will have to review it. Looks pretentious.

**Comment by jeff-zed** [comment]
> Wow. We will have to review it. Looks pretentious.

Thanks? I am working on extending it with `stressng`.

File: tests/memory-testing/scripts/maximum-memory-test.sh [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+
+test_memory_pressure() {
+    echo "=== Memory Pressure Test ==="
+    echo "Testing behavior under memory pressure conditions"
+    
+    local base_allocation=$(echo "$AVAILABLE_RAM_MB * 0.7" | bc | cut -d. -f1)
+    echo "Step 1: Deploy base container with ${base_allocation}MB"
+    test_single_container $base_allocation "PRESSURE_BASE"
+    
+    echo "Step 2: Add additional containers while base is running"
```
Comment (eriknordmark) [inline tests/memory-testing/scripts/maximum-memory-test.sh:296]: How does the add anything? test_single_containers does a cleanup_containers at the end.

**Review by eriknordmark — CHANGES_REQUESTED** [review]
Is this ready for review and validation? It is marked as WIP and draft, and when I follow the instructions (after correcting the missing "/" in the path names it doesn't find the application to run. See below.

One thing I wanted to understand is what disk and network I/O stress tests you are running from the VM. From what I can tell you merely boot different size VMs but the issue is explicit about running such stress tests.

Failure to run basic test:

nordmark@bobo:~/lf-edge/issue-4247-memory-tests/eve$ ./tests/memory-testing/scripts/simple-memory-test.sh
=== Simple EVE Memory Allocation Test ===
Results: ../results/simple-memory-20250702_121558

=== Initial System State ===
✔ Adam status: container with name eden_adam is running
	Adam is expected at https://172.16.0.148:3333
	For local Adam you can run 'docker logs eden_adam' to see logs
✔ Registry status: container with name eden_registry is running
	Registry is expected at https://172.16.0.148:5050
	For local registry you can run 'docker logs eden_registry' to see logs
✔ Redis status: container with name eden_redis is running
	Redis is expected at 172.16.0.148:6379
	For local Redis you can run 'docker logs eden_redis' to see logs
✔ EServer process status: container with name eden_eserver is running
	EServer is expected at http://172.16.0.148:8888 from EVE
	For local EServer you can run 'docker logs eden_eserver' to see logs

--- context: default ---
EVE state: registered

✔ EVE REMOTE IPs: 192.168.0.10; fe80::6c92:de61:d19f:f725; 192.168.0.11; fe80::9f4c:216d:c141:b45
	Last info received time: 2025-07-02 12:11:27 -0700 PDT
✔ EVE memory: 749 MB/2.8 GB
✔ EVE on Qemu status: running with pid 784376
	Logs for local EVE at: /home/nordmark/main/eden/dist/default-eve.log
✔ EVE Request IP: 172.16.0.148
------

=== Testing: light (128 MB) ===
Description: Light memory allocation test
INFO[0000] not found app with name simple-memory-test-light 
Pre-test EVE memory: memory:
Deploying container with 128MB memory...
config changed, to see config run 'eden controller edge-node get-config'
INFO[0002] deploy pod simple-memory-test-light with docker://ubuntu:20.04 request sent 
✅ Deployment command successful
Waiting... (0/120) Status: UNKNOWN

**Comment by rene** [comment]
Closing due to inactivity. Anyone taking this over, please, start from a fresh PR.
